shpc/main/registry/github.py

```diff
@@ -0,0 +1,174 @@
+    """
+    A remote result provides courtesy functions for interacting with
+    a container yaml recipe on GitHub.
+    """
+
+    def __init__(self, module, spec, load=True, config=None):
+        self.module = module
+        self._spec = spec
+        self._config = config
+        if load:
+            self.load(spec)
+
+    def load(self, spec):
+        """
+        Load the settings file into the settings object
+        """
+        if "config" not in spec:
+            logger.exit("Cannot find required 'config' key in remote metadata.")
+        self.package_file = spec["config_url"]
+        self._config = spec["config"]
+
+    def get_overrides(self, tag):
+        """
+        Load a filesystem based override file
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+    def save(self, package_file):
+        raise ValueError("Remote save to a GitHub registry is not supported.")
+
+    def load_wrapper_script(self, container_tech, script):
+        """
+        Get the content of a wrapper script, if it exists.
+        """
+        wrapper_script = self.find_wrapper_script(container_tech, script)
+
+        if wrapper_script:
+            url = os.path.join(self.dirname, wrapper_script)
+            response = requests.get(url)
+            if response.status_code != 200:
+                logger.warning(
+                    "Could not find wrapper script %s in remote registry." % script
+                )
+            return response.text
+
+    def override_exists(self, tag):
+        """
+        Check if an override exists.
+        """
+        logger.exit(
+            "We don't have support for remote overrides yet! Please open an issue!"
+        )
+
+
+class GitHub(Provider):
+    """
+    Currently a GitHub provider is only needed for on-demand upgrade.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.tag = kwargs.get("tag")
+
+        # Cache of remote container metadata
+        self._cache = {}
+
+        # E.g., subdirectory with registry files
+        self.subdir = kwargs.get("subdir")
+        super().__init__(*args, **kwargs)
+        self._url = self.source
+
+    def exists(self, name):
+        """
+        Determine if a module exists in the registry.
+        """
+        dirname = self.source
+        if self.subdir:
+            dirname = os.path.join(dirname, self.subdir)
+        return os.path.exists(os.path.join(dirname, name))
+
+    @classmethod
+    def matches(cls, source):
+        return cls.__name__.lower() in source and source.startswith("http")
+
+    def clone(self, tmpdir=None):
+        """
+        Clone the known source URL to a temporary directory
+        """
+        tmpdir = tmpdir or shpc.utils.get_tmpdir()
+
+        cmd = ["git", "clone", "--depth", "1"]
+        if self.tag:
+            cmd += ["-b", self.tag]
+        cmd += [self._url, tmpdir]
+        self.source = tmpdir
```

Comment:
This bit here could work with _any_ git transport mechanism (e.g. incl. SSH) and any git server (e.g. incl. GitLab), which is great ! What do you think it would take to remove the https+github restriction ?

Reply:
The ideal bit about GitHub is that it has GitHub pages to provide a static "API" that makes the interactions more efficient (e.g., I can get contents without clone). My suggestion is to start with GitLab and put up a registry with this same automation and pages, and then I can write the controller. Then we can tackle the others, and we just won't have this ability to look before clone. Would that work?